include/exadg/fluid_structure_interaction_precice/coupling_interface.h

```diff
@@ -0,0 +1,224 @@
+
+  /// The MatrixFree object (preCICE can only handle double precision)
+  std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> mf_data;
+
+  /// public precice solverinterface
+  std::shared_ptr<precice::SolverInterface> precice;
+
+  /// Configuration parameters
+  const std::string mesh_name;
+  int               mesh_id = -1;
+  // Map between data ID (preCICE) and the data name
+  std::map<std::string, int> read_data_map;
+  std::map<std::string, int> write_data_map;
+
+  const types::boundary_id dealii_boundary_interface_id;
+
+  WriteDataType write_data_type = WriteDataType::undefined;
+
+  virtual std::string
+  get_interface_type() const = 0;
+};
+
+
+
+template<int dim, int data_dim, typename VectorizedArrayType>
+CouplingInterface<dim, data_dim, VectorizedArrayType>::CouplingInterface(
+  std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> data,
+  std::shared_ptr<precice::SolverInterface>                           precice,
+  std::string                                                         mesh_name,
+  const types::boundary_id                                            interface_id)
+  : mf_data(data),
+    precice(precice),
+    mesh_name(mesh_name),
+    dealii_boundary_interface_id(interface_id)
+{
+  Assert(data.get() != nullptr, ExcNotInitialized());
+  Assert(precice.get() != nullptr, ExcNotInitialized());
+
+  // Ask preCICE already in the constructor for the IDs
+  mesh_id = precice->getMeshID(mesh_name);
+}
+
+
+template<int dim, int data_dim, typename VectorizedArrayType>
+void
+CouplingInterface<dim, data_dim, VectorizedArrayType>::add_read_data(
+  const std::string & read_data_name_)
+{
+  Assert(mesh_id != -1, ExcNotInitialized());
+  const int read_data_id = precice->getDataID(read_data_name_, mesh_id);
+  read_data_map.insert({read_data_name_, read_data_id});
+}
+
+
+
+template<int dim, int data_dim, typename VectorizedArrayType>
+void
+CouplingInterface<dim, data_dim, VectorizedArrayType>::add_write_data(
+  const std::string & write_data_name_,
+  const std::string & write_data_specification)
+{
+  Assert(mesh_id != -1, ExcNotInitialized());
+  const int write_data_id = precice->getDataID(write_data_name_, mesh_id);
+  write_data_map.insert({write_data_name_, write_data_id});
+
+  if(write_data_specification == "values_on_dofs")
+    write_data_type = WriteDataType::values_on_dofs;
+  else if(write_data_specification == "values_on_other_mesh")
+    write_data_type = WriteDataType::values_on_other_mesh;
+  else if(write_data_specification == "gradients_on_other_mesh")
+    write_data_type = WriteDataType::gradients_on_other_mesh;
+  else if(write_data_specification == "values_on_quads")
+    write_data_type = WriteDataType::values_on_quads;
+  else if(write_data_specification == "normal_gradients_on_quads")
+    write_data_type = WriteDataType::normal_gradients_on_quads;
+  else
+    AssertThrow(false, ExcMessage("Unknwon write data type."));
```

Comment:
I would introduce a free function `string_to_enum()` as in other places of ExaDG.